pandas-stubs/core/algorithms.pyi

```diff
@@ -5,65 +5,75 @@ from typing import (
 )
 
 import numpy as np
-from pandas import (
-    Categorical,
-    CategoricalIndex,
-    Index,
-    IntervalIndex,
-    PeriodIndex,
-    Series,
-)
+from numpy import typing as npt
 from pandas.api.extensions import ExtensionArray
+from pandas.core.arrays.categorical import Categorical
+from pandas.core.indexes.base import Index
+from pandas.core.indexes.category import CategoricalIndex
+from pandas.core.indexes.datetimes import DatetimeIndex
+from pandas.core.indexes.interval import IntervalIndex
+from pandas.core.indexes.period import PeriodIndex
+from pandas.core.series import Series
 
 from pandas._typing import (
+    T_EXTENSION_ARRAY,
     AnyArrayLike,
+    GenericT,
     IntervalT,
     TakeIndexer,
     np_1darray,
+    np_ndarray,
 )
 
 # These are type: ignored because the Index types overlap due to inheritance but indices
 # with extension types return the same type while standard type return ndarray
-
-@overload
-def unique(  # pyright: ignore[reportOverlappingOverload]
-    values: PeriodIndex,
-) -> PeriodIndex: ...
 @overload
-def unique(values: CategoricalIndex) -> CategoricalIndex: ...  # type: ignore[overload-overlap]
+def unique(values: CategoricalIndex) -> CategoricalIndex: ...
 @overload
 def unique(values: IntervalIndex[IntervalT]) -> IntervalIndex[IntervalT]: ...
 @overload
-def unique(values: Index) -> np.ndarray: ...
+def unique(values: PeriodIndex) -> PeriodIndex: ...
+@overload
+def unique(values: DatetimeIndex) -> np_1darray[np.datetime64] | DatetimeIndex: ...
+@overload
+def unique(values: Index) -> np_1darray[Any] | Index: ...
```

Comment:
I'm surprised that these are not always returning the corresponding `Index` type, i.e., I don't see how they could return a numpy array